--- src/lerobot/scripts/rl/gym_manipulator.py	2025-08-29 16:06:24.009032079 +0000
+++ matchbox/lerobot/scripts/rl/gym_manipulator.py	2025-08-29 15:03:37.396837963 +0000
@@ -39,38 +39,37 @@ Example:
 import logging
 import time
 from collections import deque
-from collections.abc import Sequence
 from threading import Lock
-from typing import Annotated, Any
+from typing import Annotated, Any, Sequence
 
 import gymnasium as gym
 import numpy as np
 import torch
 import torchvision.transforms.functional as F  # noqa: N812
 
-from lerobot.cameras import opencv  # noqa: F401
-from lerobot.configs import parser
-from lerobot.envs.configs import EnvConfig
-from lerobot.envs.utils import preprocess_observation
-from lerobot.model.kinematics import RobotKinematics
-from lerobot.robots import (  # noqa: F401
+from lerobot.common.cameras import opencv  # noqa: F401
+from lerobot.common.envs.configs import EnvConfig
+from lerobot.common.envs.utils import preprocess_observation
+from lerobot.common.model.kinematics import RobotKinematics
+from lerobot.common.robots import (  # noqa: F401
     RobotConfig,
     make_robot_from_config,
     so100_follower,
-    so101_follower
+    so101_follower,
 )
-from lerobot.teleoperators import (
+from lerobot.common.teleoperators import (
     gamepad,  # noqa: F401
     keyboard,  # noqa: F401
     make_teleoperator_from_config,
     so101_leader,  # noqa: F401
 )
-from lerobot.teleoperators.gamepad.teleop_gamepad import GamepadTeleop
-from lerobot.teleoperators.keyboard.teleop_keyboard import KeyboardEndEffectorTeleop
-from lerobot.utils.robot_utils import busy_wait
-from lerobot.utils.utils import log_say
+from lerobot.common.teleoperators.gamepad.teleop_gamepad import GamepadTeleop
+from lerobot.common.teleoperators.keyboard.teleop_keyboard import KeyboardEndEffectorTeleop
+from lerobot.common.utils.robot_utils import busy_wait
+from lerobot.common.utils.utils import log_say
+from lerobot.configs import parser
 
-logging.basicConfig(level=logging.INFO)
+logging.basicConfig(level=logging.INFO)  # Changed from INFO to DEBUG
 
 
 def reset_follower_position(robot_arm, target_position):
@@ -227,6 +226,7 @@ class RobotEnv(gym.Env):
     def __init__(
         self,
         robot,
+        action_space_type: str,
         use_gripper: bool = False,
         display_cameras: bool = False,
     ):
@@ -238,12 +238,14 @@ class RobotEnv(gym.Env):
 
         Args:
             robot: The robot interface object used to connect and interact with the physical robot.
+            action_space_type: The control mode, e.g., 'end_effector' or 'joint'.
             display_cameras: If True, the robot's camera feeds will be displayed during execution.
         """
         super().__init__()
 
         self.robot = robot
         self.display_cameras = display_cameras
+        self.action_space_type = action_space_type
 
         # Connect to the robot if not already connected.
         if not self.robot.is_connected:
@@ -256,19 +258,20 @@ class RobotEnv(gym.Env):
         self._joint_names = [f"{key}.pos" for key in self.robot.bus.motors]
         self._image_keys = self.robot.cameras.keys()
 
-        self.current_observation = None
+        # Read initial joint positions using the bus
+        self.current_joint_positions = self._get_observation()["agent_pos"]
 
         self.use_gripper = use_gripper
 
         self._setup_spaces()
 
-    def _get_observation(self) -> dict[str, np.ndarray]:
+    def _get_observation(self) -> np.ndarray:
         """Helper to convert a dictionary from bus.sync_read to an ordered numpy array."""
         obs_dict = self.robot.get_observation()
-        joint_positions = np.array([obs_dict[name] for name in self._joint_names])
+        joint_positions = np.array([obs_dict[name] for name in self._joint_names], dtype=np.float32)
 
         images = {key: obs_dict[key] for key in self._image_keys}
-        self.current_observation = {"agent_pos": joint_positions, "pixels": images}
+        return {"agent_pos": joint_positions, "pixels": images}
 
     def _setup_spaces(self):
         """
@@ -282,39 +285,45 @@ class RobotEnv(gym.Env):
             - The action space is defined as a Box space representing joint position commands. It is defined as relative (delta)
               or absolute, based on the configuration.
         """
-        self._get_observation()
+        example_obs = self._get_observation()
 
         observation_spaces = {}
 
         # Define observation spaces for images and other states.
-        if "pixels" in self.current_observation:
-            prefix = "observation.images"
+        if "pixels" in example_obs:
+            prefix = "observation.images" if len(example_obs["pixels"]) > 1 else "observation.image"
             observation_spaces = {
                 f"{prefix}.{key}": gym.spaces.Box(
-                    low=0, high=255, shape=self.current_observation["pixels"][key].shape, dtype=np.uint8
+                    low=0, high=255, shape=example_obs["pixels"][key].shape, dtype=np.uint8
                 )
-                for key in self.current_observation["pixels"]
+                for key in example_obs["pixels"]
             }
 
         observation_spaces["observation.state"] = gym.spaces.Box(
             low=0,
             high=10,
-            shape=self.current_observation["agent_pos"].shape,
+            shape=example_obs["agent_pos"].shape,
             dtype=np.float32,
         )
 
         self.observation_space = gym.spaces.Dict(observation_spaces)
 
         # Define the action space for joint positions along with setting an intervention flag.
-        action_dim = 3
-        bounds = {}
-        bounds["min"] = -np.ones(action_dim)
-        bounds["max"] = np.ones(action_dim)
+        if self.action_space_type == "joint":
+            action_dim = len(self.robot.bus.motors)
+            bounds = {"min": -np.ones(action_dim), "max": np.ones(action_dim)}
+        else:  # end_effector
+            action_dim = 3
+            bounds = {"min": -np.ones(action_dim), "max": np.ones(action_dim)}
 
         if self.use_gripper:
-            action_dim += 1
-            bounds["min"] = np.concatenate([bounds["min"], [0]])
-            bounds["max"] = np.concatenate([bounds["max"], [2]])
+            if self.action_space_type == "joint":
+                bounds["min"][-1] = 0
+                bounds["max"][-1] = 2
+            else:
+                action_dim += 1
+                bounds["min"] = np.concatenate([bounds["min"], [0]])
+                bounds["max"] = np.concatenate([bounds["max"], [2]])
 
         self.action_space = gym.spaces.Box(
             low=bounds["min"],
@@ -341,12 +350,14 @@ class RobotEnv(gym.Env):
 
         self.robot.reset()
 
+        # Capture the initial observation.
+        observation = self._get_observation()
+
         # Reset episode tracking variables.
         self.current_step = 0
         self.episode_data = None
-        self.current_observation = None
-        self._get_observation()
-        return self.current_observation, {"is_intervention": False}
+
+        return observation, {"is_intervention": False}
 
     def step(self, action) -> tuple[dict[str, np.ndarray], float, bool, bool, dict[str, Any]]:
         """
@@ -366,14 +377,18 @@ class RobotEnv(gym.Env):
                 - truncated (bool): True if the episode was truncated (e.g., time constraints).
                 - info (dict): Additional debugging information including intervention status.
         """
-        action_dict = {"delta_x": action[0], "delta_y": action[1], "delta_z": action[2]}
+        self.current_joint_positions = self._get_observation()["agent_pos"]
 
-        # 1.0 action corresponds to no-op action
-        action_dict["gripper"] = action[3] if self.use_gripper else 1.0
+        if self.action_space_type == "joint":
+            # In joint space mode, the action from the wrapper is an array of absolute joint positions.
+            action_dict = {f"{name}": action[i] for i, name in enumerate(self._joint_names)}
+        else:
+            action_dict = {"delta_x": action[0], "delta_y": action[1], "delta_z": action[2]}
 
-        self.robot.send_action(action_dict)
+        # The last element of the action array is always the gripper. 1.0 corresponds to no-op action
+        action_dict["gripper"] = action[-1] if self.use_gripper else 1.0
 
-        self._get_observation()
+        self.robot.send_action(action_dict)
 
         if self.display_cameras:
             self.render()
@@ -385,7 +400,7 @@ class RobotEnv(gym.Env):
         truncated = False
 
         return (
-            self.current_observation,
+            self._get_observation(),
             reward,
             terminated,
             truncated,
@@ -398,10 +413,11 @@ class RobotEnv(gym.Env):
         """
         import cv2
 
-        image_keys = [key for key in self.current_observation if "image" in key]
+        observation = self._get_observation()
+        image_keys = [key for key in observation if "image" in key]
 
         for key in image_keys:
-            cv2.imshow(key, cv2.cvtColor(self.current_observation[key].numpy(), cv2.COLOR_RGB2BGR))
+            cv2.imshow(key, cv2.cvtColor(observation[key].numpy(), cv2.COLOR_RGB2BGR))
             cv2.waitKey(1)
 
     def close(self):
@@ -518,10 +534,7 @@ class AddCurrentToObservation(gym.Observ
         Returns:
             The modified observation with current values.
         """
-        present_current_dict = self.env.unwrapped.robot.bus.sync_read("Present_Current")
-        present_current_observation = np.array(
-            [present_current_dict[name] for name in self.env.unwrapped.robot.bus.motors]
-        )
+        present_current_observation = self.unwrapped._get_observation()["agent_pos"]
         observation["agent_pos"] = np.concatenate(
             [observation["agent_pos"], present_current_observation], axis=-1
         )
@@ -571,6 +584,7 @@ class RewardWrapper(gym.Wrapper):
                 if self.reward_classifier is not None
                 else 0.0
             )
+
         info["Reward classifier frequency"] = 1 / (time.perf_counter() - start_time)
 
         reward = 0.0
@@ -849,10 +863,20 @@ class ResetWrapper(gym.Wrapper):
             log_say("Reset the environment done.", play_sounds=True)
 
             if hasattr(self.env, "robot_leader"):
+                is_intervention = self.env._check_intervention() if hasattr(self.env, "_check_intervention") else False
+
                 self.env.robot_leader.bus.sync_write("Torque_Enable", 1)
+                if hasattr(self.env, "leader_torque_enabled"):
+                    self.env.leader_torque_enabled = True
+
                 log_say("Reset the leader robot.", play_sounds=True)
                 reset_follower_position(self.env.robot_leader, self.reset_pose)
                 log_say("Reset the leader robot done.", play_sounds=True)
+
+                if is_intervention:
+                    self.env.robot_leader.bus.sync_write("Torque_Enable", 0)
+                    if hasattr(self.env, "leader_torque_enabled"):
+                        self.env.leader_torque_enabled = False
         else:
             log_say(
                 f"Manually reset the environment for {self.reset_time_s} seconds.",
@@ -1091,10 +1115,13 @@ class EEObservationWrapper(gym.Observati
             dtype=np.float32,
         )
 
-        self.kinematics = RobotKinematics(
-            urdf_path=env.unwrapped.robot.config.urdf_path,
-            target_frame_name=env.unwrapped.robot.config.target_frame_name,
-        )
+        # Initialize kinematics instance for the appropriate robot type
+        robot_type = getattr(env.unwrapped.robot.config, "robot_type", "so101")
+        if "so100" in robot_type or "so101" in robot_type:
+            # Note to be compatible with the rest of the codebase,
+            # we are using the new calibration method for so101 and so100
+            robot_type = "so_new_calibration"
+        self.kinematics = RobotKinematics(robot_type)
 
     def observation(self, observation):
         """
@@ -1106,9 +1133,9 @@ class EEObservationWrapper(gym.Observati
         Returns:
             Enhanced observation with end-effector pose information.
         """
-        current_joint_pos = self.unwrapped.current_observation["agent_pos"]
+        current_joint_pos = self.unwrapped._get_observation()["agent_pos"]
 
-        current_ee_pos = self.kinematics.forward_kinematics(current_joint_pos)[:3, 3]
+        current_ee_pos = self.kinematics.forward_kinematics(current_joint_pos, frame="gripper_tip")[:3, 3]
         observation["agent_pos"] = np.concatenate([observation["agent_pos"], current_ee_pos], -1)
         return observation
 
@@ -1130,6 +1157,7 @@ class BaseLeaderControlWrapper(gym.Wrapp
         self,
         env,
         teleop_device,
+        action_space_type: str,
         end_effector_step_sizes,
         use_geared_leader_arm: bool = False,
         use_gripper=False,
@@ -1140,6 +1168,7 @@ class BaseLeaderControlWrapper(gym.Wrapp
         Args:
             env: The environment to wrap.
             teleop_device: The teleoperation device.
+            action_space_type: The control mode, e.g., 'end_effector' or 'joint'.
             use_geared_leader_arm: Whether to use a geared leader arm setup.
             use_gripper: Whether to include gripper control.
         """
@@ -1148,17 +1177,22 @@ class BaseLeaderControlWrapper(gym.Wrapp
         self.robot_follower = env.unwrapped.robot
         self.use_geared_leader_arm = use_geared_leader_arm
         self.use_gripper: bool = use_gripper
-        self.end_effector_step_sizes = np.array(list(end_effector_step_sizes.values()))
+        self.action_space_type = action_space_type
+        self.end_effector_step_sizes = (
+            np.array(list(end_effector_step_sizes.values())) if end_effector_step_sizes is not None else None
+        )
 
         # Set up keyboard event tracking
         self._init_keyboard_events()
         self.event_lock = Lock()  # Thread-safe access to events
 
         # Initialize robot control
-        self.kinematics = RobotKinematics(
-            urdf_path=env.unwrapped.robot.config.urdf_path,
-            target_frame_name=env.unwrapped.robot.config.target_frame_name,
-        )
+        robot_type = getattr(env.unwrapped.robot.config, "robot_type", "so101")
+        if "so100" in robot_type or "so101" in robot_type:
+            # Note to be compatible with the rest of the codebase,
+            # we are using the new calibration method for so101 and so100
+            robot_type = "so_new_calibration"
+        self.kinematics = RobotKinematics(robot_type)
         self.leader_torque_enabled = True
         self.prev_leader_gripper = None
 
@@ -1256,37 +1290,48 @@ class BaseLeaderControlWrapper(gym.Wrapp
         leader_pos_dict = self.robot_leader.bus.sync_read("Present_Position")
         follower_pos_dict = self.robot_follower.bus.sync_read("Present_Position")
 
-        leader_pos = np.array([leader_pos_dict[name] for name in leader_pos_dict])
-        follower_pos = np.array([follower_pos_dict[name] for name in follower_pos_dict])
+        leader_pos = np.array(
+            [leader_pos_dict[name] for name in self.robot_leader.bus.motors], dtype=np.float32
+        )
+        follower_pos = np.array(
+            [follower_pos_dict[name] for name in self.robot_follower.bus.motors], dtype=np.float32
+        )
 
         self.leader_tracking_error_queue.append(np.linalg.norm(follower_pos[:-1] - leader_pos[:-1]))
 
-        # [:3, 3] Last column of the transformation matrix corresponds to the xyz translation
-        leader_ee = self.kinematics.forward_kinematics(leader_pos)[:3, 3]
-        follower_ee = self.kinematics.forward_kinematics(follower_pos)[:3, 3]
+        if self.action_space_type == "end_effector":
+            # [:3, 3] Last column of the transformation matrix corresponds to the xyz translation
+            leader_ee = self.kinematics.forward_kinematics(leader_pos, frame="gripper_tip")[:3, 3]
+            follower_ee = self.kinematics.forward_kinematics(follower_pos, frame="gripper_tip")[:3, 3]
 
-        action = np.clip(leader_ee - follower_ee, -self.end_effector_step_sizes, self.end_effector_step_sizes)
-        # Normalize the action to the range [-1, 1]
-        action = action / self.end_effector_step_sizes
+            action = np.clip(leader_ee - follower_ee, -self.end_effector_step_sizes, self.end_effector_step_sizes)
+            # Normalize the action to the range [-1, 1]
+            action = action / self.end_effector_step_sizes
+        else:
+            # For joint space control, the action is the leader's absolute joint positions.
+            action = leader_pos[:-1]
 
         if self.use_gripper:
-            if self.prev_leader_gripper is None:
-                self.prev_leader_gripper = np.clip(
-                    leader_pos[-1], 0, self.robot_follower.config.max_gripper_pos
-                )
+            if self.action_space_type == "end_effector":
+                if self.prev_leader_gripper is None:
+                    self.prev_leader_gripper = np.clip(
+                        leader_pos[-1], 0, self.robot_follower.config.max_gripper_pos
+                    )
 
-            # Get gripper action delta based on leader pose
-            leader_gripper = leader_pos[-1]
-            gripper_delta = leader_gripper - self.prev_leader_gripper
+                # Get gripper action delta based on leader pose
+                leader_gripper = leader_pos[-1]
+                gripper_delta = leader_gripper - self.prev_leader_gripper
 
-            # Normalize by max angle and quantize to {0,1,2}
-            normalized_delta = gripper_delta / self.robot_follower.config.max_gripper_pos
-            if normalized_delta >= 0.3:
-                gripper_action = 2
-            elif normalized_delta <= 0.1:
-                gripper_action = 0
+                # Normalize by max angle and quantize to {0,1,2}
+                normalized_delta = gripper_delta / self.robot_follower.config.max_gripper_pos
+                if normalized_delta >= 0.3:
+                    gripper_action = 2
+                elif normalized_delta <= -0.3:
+                    gripper_action = 0
+                else:
+                    gripper_action = 1
             else:
-                gripper_action = 1
+                gripper_action = leader_pos[-1]
 
             action = np.append(action, gripper_action)
 
@@ -1337,9 +1382,6 @@ class BaseLeaderControlWrapper(gym.Wrapp
         # NOTE:
         obs, reward, terminated, truncated, info = self.env.step(action)
 
-        if isinstance(action, np.ndarray):
-            action = torch.from_numpy(action)
-
         # Add intervention info
         info["is_intervention"] = is_intervention
         info["action_intervention"] = action
@@ -1394,6 +1436,10 @@ class GearedLeaderControlWrapper(BaseLea
     of human intervention mode with keyboard controls.
     """
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.is_manual_intervention_active = False
+
     def _init_keyboard_events(self):
         """
         Initialize keyboard events including human intervention flag.
@@ -1402,7 +1448,6 @@ class GearedLeaderControlWrapper(BaseLea
         intervention state toggled by keyboard.
         """
         super()._init_keyboard_events()
-        self.keyboard_events["human_intervention_step"] = False
 
     def _handle_key_press(self, key, keyboard_device):
         """
@@ -1416,15 +1461,14 @@ class GearedLeaderControlWrapper(BaseLea
         """
         super()._handle_key_press(key, keyboard_device)
         if key == keyboard_device.Key.space:
-            if not self.keyboard_events["human_intervention_step"]:
+            self.is_manual_intervention_active = not self.is_manual_intervention_active
+            if self.is_manual_intervention_active:
                 logging.info(
                     "Space key pressed. Human intervention required.\n"
                     "Place the leader in similar pose to the follower and press space again."
                 )
-                self.keyboard_events["human_intervention_step"] = True
                 log_say("Human intervention step.", play_sounds=True)
             else:
-                self.keyboard_events["human_intervention_step"] = False
                 logging.info("Space key pressed for a second time.\nContinuing with policy actions.")
                 log_say("Continuing with policy actions.", play_sounds=True)
 
@@ -1436,7 +1480,7 @@ class GearedLeaderControlWrapper(BaseLea
             Boolean indicating whether intervention mode is active.
         """
         return True
-        return self.keyboard_events["human_intervention_step"]
+        return self.is_manual_intervention_active
 
 
 class GearedLeaderAutomaticControlWrapper(BaseLeaderControlWrapper):
@@ -1451,6 +1495,7 @@ class GearedLeaderAutomaticControlWrappe
         self,
         env,
         teleop_device,
+        action_space_type,
         end_effector_step_sizes,
         use_gripper=False,
         intervention_threshold=10.0,
@@ -1462,12 +1507,19 @@ class GearedLeaderAutomaticControlWrappe
         Args:
             env: The environment to wrap.
             teleop_device: The teleoperation device.
+            action_space_type: The control mode, e.g., 'end_effector' or 'joint'.
             use_gripper: Whether to include gripper control.
             intervention_threshold: Error threshold to trigger intervention.
             release_threshold: Error threshold to release intervention.
             queue_size: Number of error measurements to track for smoothing.
         """
-        super().__init__(env, teleop_device, end_effector_step_sizes, use_gripper=use_gripper)
+        super().__init__(
+            env,
+            teleop_device,
+            action_space_type,
+            end_effector_step_sizes=end_effector_step_sizes,
+            use_gripper=use_gripper,
+        )
 
         # Error tracking parameters
         self.intervention_threshold = intervention_threshold  # Threshold to trigger intervention
@@ -1877,12 +1929,14 @@ def make_robot_env(cfg: EnvConfig) -> gy
     if cfg.robot is None:
         raise ValueError("RobotConfig (cfg.robot) must be provided for gym_manipulator environment.")
     robot = make_robot_from_config(cfg.robot)
+
     teleop_device = make_teleoperator_from_config(cfg.teleop)
     teleop_device.connect()
 
     # Create base environment
     env = RobotEnv(
         robot=robot,
+        action_space_type=cfg.wrapper.action_space,
         use_gripper=cfg.wrapper.use_gripper,
         display_cameras=cfg.wrapper.display_cameras if cfg.wrapper else False,
     )
@@ -1918,41 +1972,45 @@ def make_robot_env(cfg: EnvConfig) -> gy
         )
 
     # Control mode specific wrappers
-    control_mode = cfg.wrapper.control_mode
-    if control_mode == "gamepad":
-        assert isinstance(teleop_device, GamepadTeleop), (
-            "teleop_device must be an instance of GamepadTeleop for gamepad control mode"
-        )
+    teleop_device_type = cfg.teleop.type
+    if teleop_device_type == "gamepad":
+        assert isinstance(
+            teleop_device, GamepadTeleop
+        ), "teleop_device must be an instance of GamepadTeleop for gamepad control mode"
         env = GamepadControlWrapper(
             env=env,
             teleop_device=teleop_device,
             use_gripper=cfg.wrapper.use_gripper,
         )
-    elif control_mode == "keyboard_ee":
-        assert isinstance(teleop_device, KeyboardEndEffectorTeleop), (
-            "teleop_device must be an instance of KeyboardEndEffectorTeleop for keyboard control mode"
-        )
+    elif teleop_device_type == "keyboard_ee":
+        assert isinstance(
+            teleop_device, KeyboardEndEffectorTeleop
+        ), "teleop_device must be an instance of KeyboardEndEffectorTeleop for keyboard control mode"
         env = KeyboardControlWrapper(
             env=env,
             teleop_device=teleop_device,
             use_gripper=cfg.wrapper.use_gripper,
         )
-    elif control_mode == "leader":
+    elif "leader" in teleop_device_type and not "automatic" in teleop_device_type:
         env = GearedLeaderControlWrapper(
             env=env,
             teleop_device=teleop_device,
-            end_effector_step_sizes=cfg.robot.end_effector_step_sizes,
+            action_space_type=cfg.wrapper.action_space,
+            end_effector_step_sizes=(
+                cfg.robot.end_effector_step_sizes if hasattr(cfg.robot, "end_effector_step_sizes") else None
+            ),
             use_gripper=cfg.wrapper.use_gripper,
         )
-    elif control_mode == "leader_automatic":
+    elif "leader_automatic" in teleop_device_type:
         env = GearedLeaderAutomaticControlWrapper(
             env=env,
             teleop_device=teleop_device,
+            action_space_type=cfg.wrapper.action_space,
             end_effector_step_sizes=cfg.robot.end_effector_step_sizes,
             use_gripper=cfg.wrapper.use_gripper,
         )
     else:
-        raise ValueError(f"Invalid control mode: {control_mode}")
+        raise ValueError(f"Invalid teleop device type: {teleop_device_type}")
 
     env = ResetWrapper(
         env=env,
@@ -1979,7 +2037,7 @@ def init_reward_classifier(cfg):
     if cfg.reward_classifier_pretrained_path is None:
         return None
 
-    from lerobot.policies.sac.reward_model.modeling_classifier import Classifier
+    from lerobot.common.policies.sac.reward_model.modeling_classifier import Classifier
 
     # Get device from config or default to CUDA
     device = getattr(cfg, "device", "cpu")
@@ -2022,14 +2080,20 @@ def record_dataset(env, policy, cfg):
                                   a success (reward=1) is detected. This helps collect
                                   more positive examples for reward classifier training.
     """
-    from lerobot.datasets.lerobot_dataset import LeRobotDataset
+    from lerobot.common.datasets.lerobot_dataset import LeRobotDataset
 
     # Setup initial action (zero action if using teleop)
     action = env.action_space.sample() * 0.0
 
-    action_names = ["delta_x_ee", "delta_y_ee", "delta_z_ee"]
-    if cfg.wrapper.use_gripper:
-        action_names.append("gripper_delta")
+    if cfg.wrapper.action_space == "end_effector":
+        action_names = ["delta_x_ee", "delta_y_ee", "delta_z_ee"]
+        if cfg.wrapper.use_gripper:
+            action_names.append("gripper_delta")
+    else:  # joint space
+        # get joint names from env
+        action_names = env.unwrapped._joint_names
+        if not cfg.wrapper.use_gripper:
+            action_names = action_names[:-1]
 
     # Configure dataset features based on environment spaces
     features = {
@@ -2040,7 +2104,7 @@ def record_dataset(env, policy, cfg):
         },
         "action": {
             "dtype": "float32",
-            "shape": (len(action_names),),
+            "shape": env.action_space.shape,
             "names": action_names,
         },
         "next.reward": {"dtype": "float32", "shape": (1,), "names": None},
@@ -2055,10 +2119,12 @@ def record_dataset(env, policy, cfg):
     # Add image features
     for key in env.observation_space:
         if "image" in key:
+            # The observation space is CHW, but dataset expects HWC for video.
+            c, h, w = env.observation_space[key].shape
             features[key] = {
                 "dtype": "video",
-                "shape": env.observation_space[key].shape,
-                "names": ["channels", "height", "width"],
+                "shape": (h, w, c),
+                "names": ["height", "width", "channels"],
             }
 
     # Create dataset
@@ -2100,9 +2166,15 @@ def record_dataset(env, policy, cfg):
                 break
 
             # For teleop, get action from intervention
-            recorded_action = {
-                "action": info["action_intervention"].cpu().squeeze(0).float() if policy is None else action
-            }
+            if policy is None:
+                action = (
+                    info["action_intervention"].cpu().squeeze(0).float()
+                    if torch.is_tensor(info["action_intervention"])
+                    else torch.from_numpy(info["action_intervention"])
+                )
+            else:
+                action = action
+            recorded_action = {"action": action.to(torch.float32) if torch.is_tensor(action) else action.astype(np.float32)}
 
             # Process observation for dataset
             obs_processed = {k: v.cpu().squeeze(0).float() for k, v in obs.items()}
@@ -2176,7 +2248,7 @@ def replay_episode(env, cfg):
             - dataset_root: Local root directory for dataset
             - episode: Episode ID to replay
     """
-    from lerobot.datasets.lerobot_dataset import LeRobotDataset
+    from lerobot.common.datasets.lerobot_dataset import LeRobotDataset
 
     dataset = LeRobotDataset(cfg.repo_id, root=cfg.dataset_root, episodes=[cfg.episode])
     env.reset()
@@ -2209,7 +2281,7 @@ def main(cfg: EnvConfig):
     if cfg.mode == "record":
         policy = None
         if cfg.pretrained_policy_name_or_path is not None:
-            from lerobot.policies.sac.modeling_sac import SACPolicy
+            from lerobot.common.policies.sac.modeling_sac import SACPolicy
 
             policy = SACPolicy.from_pretrained(cfg.pretrained_policy_name_or_path)
             policy.to(cfg.device)
